TodoListWebApp/Extensions/AzureAdAuthenticationBuilderExtensions.cs

```diff
@@ -31,16 +33,56 @@ public ConfigureAzureOptions(IOptions<AzureAdOptions> azureOptions)
             public void Configure(string name, OpenIdConnectOptions options)
             {
                 options.ClientId = _azureOptions.ClientId;
-                options.Authority = $"{_azureOptions.Instance}{_azureOptions.TenantId}";
+                options.Authority = _azureOptions.Authority;
                 options.UseTokenLifetime = true;
                 options.CallbackPath = _azureOptions.CallbackPath;
                 options.RequireHttpsMetadata = false;
+                options.ClientSecret = _azureOptions.ClientSecret;
+                options.Resource = "https://graph.windows.net"; // AAD graph
+
+                // Without overriding the response type (which by default is id_token), the OnAuthorizationCodeReceived event is not called.
+                // but instead OnTokenValidated event is called. Here we request both so that OnTokenValidated is called first which 
+                // ensures that context.Principal has a non-null value when OnAuthorizeationCodeReceived is called
+                options.ResponseType = "id_token code";
```

Comment:
this bit seems to be a change of behavior in ASP.NET Core 2.0. This also provide an answer to issue #15